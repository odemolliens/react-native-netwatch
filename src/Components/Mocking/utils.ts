import RNFS from 'react-native-fs';

export const FILE_PATH = RNFS.DocumentDirectoryPath + 'Netwatch.mockResponses.35435664445664';

export const mockResponses: MockResponse[] = [];

export interface MockResponse {
  url: string;
  method?: string;
  response?: string;
  headers?: string;
  statusCode?: number;
  timeout?: number;
  active?: boolean;
  date?: number;
}

export function getMockResponses() {
  return mockResponses;
}

export function clearMockResponses() {
  mockResponses.length = 0;
  RNFS.unlink(FILE_PATH);
}

export function resetMockResponses(responses: string) {
  clearMockResponses();
  mockResponses.push(...JSON.parse(responses));
  RNFS.writeFile(FILE_PATH, responses, 'utf8');
}

export function mockRequestWithResponse(mockResponse: MockResponse) {
  let found = false;
  for (let i = 0; i < mockResponses.length; i++) {
    if (mockResponses[i].url === mockResponse.url && mockResponses[i].method === mockResponse.method) {
      mockResponses[i] = {
        ...mockResponse,
        active: true,
      };
      found = true;
      break;
    }
  }
  if (!found) {
    mockResponses.push({
      ...mockResponse,
      active: true,
    });
  }
  RNFS.writeFile(FILE_PATH, JSON.stringify(mockResponses), 'utf8');
}

let originalSend: any;

export function extractURL(url: string) {
  const urlParts = url.split('?');
  return urlParts[0];
}

// istanbul ignore next
export function setupMocks() {
  RNFS.readFile(FILE_PATH, 'utf8')
    .then((mockResponsesString: string) => {
      mockResponses.push(...(JSON.parse(mockResponsesString || '[]') as MockResponse[]));
    })
    .catch(() => console.log('No mock responses found'));
  if (originalSend) {
    console.log('mockResponses already setup');
    return;
  }
  originalSend = XMLHttpRequest.prototype.send;
  try {
    XMLHttpRequest.prototype.send = function (body: any) {
      // @ts-ignore
      const mockRequest = mockResponses.find(mock => {
        // @ts-ignore
        try {
          return (
            mock.active &&
            // @ts-ignore
            mock.url === extractURL(this._url) &&
            // @ts-ignore
            (!mock._method || (mock.method && mock._method === this._method))
          );
        } catch (e) {
          return false;
        }
      });

      if (mockRequest) {
        Object.defineProperty(this, 'response', { writable: true });
        Object.defineProperty(this, 'responseText', { writable: true });
        if (mockRequest.timeout) {
          new Promise(resolve => setTimeout(resolve, (mockRequest.timeout || 0) * 1000)).then(() => {
            this?.ontimeout?.({} as any);
          });
          return;
        }
        const originalOnReadyStateChange = this.onreadystatechange;
        this.onreadystatechange = args => {
          try {
            if (mockRequest.response && this.readyState === 4) {
              // We rewrite the response if it is defined
              // @ts-ignore
              this.responseText = mockRequest.response;
              // @ts-ignore
              this.response = mockRequest.response;
            }
            if (mockRequest.statusCode) {
              // We rewrite the status code if it is defined
              // @ts-ignore
              this.status = mockRequest.statusCode;
            }
          } catch (e) {
            console.error(e);
          }
          originalOnReadyStateChange?.apply(this, [args]);
        };
      }
      originalSend.apply(this, [body]);
    };
  } catch (e) {
    console.error(e);
  }
}
