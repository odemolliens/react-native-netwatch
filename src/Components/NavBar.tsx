import { Appbar } from 'react-native-paper';
import { Platform, StyleSheet, TouchableOpacity } from "react-native";
import FeatherIcon from 'react-native-vector-icons/Feather';
import * as React from 'react';
import { ReactNode, useContext } from 'react';
import { ThemeContext } from '../Theme';

export function NavBar(props: {
  title: string;
  rightComponent?: ReactNode;
  icon?: string;
  onPressBack: (showDetails: boolean) => void;
}) {
  const theme = useContext(ThemeContext);

  return (
    <Appbar.Header style={[styles.header, { backgroundColor: theme.secondaryDarkColor }]}>
      <TouchableOpacity
        style={[styles.button, { borderLeftWidth: 0 }]}
        onPress={() => props.onPressBack(false)}
        testID="buttonBackToMainScreen"
      >
        <FeatherIcon name={props.icon || 'arrow-left'} color={theme.textColorOne} size={24} />
      </TouchableOpacity>
      <Appbar.Content color={theme.primaryColor} title={props.title} titleStyle={{ fontSize: 18 }} />
      {props.rightComponent}
    </Appbar.Header>
  );
}

const styles = StyleSheet.create({
  header: {
    paddingTop: Platform.OS === 'ios' ? 0 : 40,
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
    height: Platform.OS === 'ios' ? 60 : 100,
  },

  container: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
  },

  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
