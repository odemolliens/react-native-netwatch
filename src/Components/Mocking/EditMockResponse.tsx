import { TextInput } from 'react-native-paper';
import { Button, KeyboardAvoidingView, Platform, ScrollView, StyleSheet, Switch, Text, View } from 'react-native';
import * as React from 'react';
import { useCallback, useContext, useEffect } from 'react';
import { ThemeContext } from '../../Theme';
import { NavBar } from '../NavBar';
import { extractURL, mockRequestWithResponse, MockResponse } from './utils';

export function EditMockResponse(props: { mockResponse: MockResponse; onPressBack: () => void; update: boolean }) {
  const theme = useContext(ThemeContext);

  const [url, setUrl] = React.useState<string>();
  const [method, setMethod] = React.useState<string>(); // [GET, POST, PUT, PATCH, DELETE
  const [headers, setHeaders] = React.useState<string>();
  const [body, setBody] = React.useState<string>();
  const [code, setCode] = React.useState<string>();
  const [active, setActive] = React.useState<boolean>(true);
  const [to, setTo] = React.useState<string>();

  useEffect(() => {
    setUrl(props.mockResponse.url);
    setMethod(props.mockResponse.method);
    setHeaders(props.mockResponse.headers);
    setBody(props.mockResponse.response);
    setCode(props.mockResponse.statusCode?.toString());
    setTo(props.mockResponse.timeout?.toString());
    setActive(props.update ? !!props.mockResponse.active : true);
  }, [
    props.mockResponse.headers,
    props.mockResponse.method,
    props.mockResponse.response,
    props.mockResponse.statusCode,
    props.mockResponse.timeout,
    props.mockResponse.url,
    props.mockResponse.active,
    props.update,
  ]);

  const saveMockResponse = useCallback(() => {
    if (url) {
      mockRequestWithResponse({
        url: extractURL(url),
        method,
        headers,
        response: body,
        statusCode: code ? parseInt(code, 10) : 0,
        timeout: to ? parseInt(to, 10) : 0,
        active,
        date: Date.now(),
      });
    }
  }, [body, code, headers, method, to, url, active]);

  return (
    <KeyboardAvoidingView style={styles.container} behavior={Platform.OS === 'ios' ? 'padding' : 'height'}>
      <NavBar
        title={'Edit Response Mock'}
        onPressBack={props.onPressBack}
        rightComponent={<Button title={'Save'} onPress={saveMockResponse} />}
      />

      <ScrollView
        keyboardDismissMode={'on-drag'}
        style={{ backgroundColor: theme.secondaryColor }}
        contentContainerStyle={styles.scrollView}
      >
        <View style={{ flex: 1, width: '100%' }}>
          <View style={{ margin: 15 }}>
            <Text style={{ color: theme.textColorOne }}>Leave blank entries to keep unchanged.</Text>
          </View>
          <View style={{ flexDirection: 'row', margin: 15 }}>
            <Text style={{ flex: 1, color: theme.textColorOne, fontSize: 14, marginBottom: 10 }}>
              Mock response active
            </Text>
            <Switch value={active} onValueChange={setActive} />
          </View>

          <TextInput mode={'flat'} style={styles.textInput} label={'URL'} onChangeText={setUrl} value={url} />
          <TextInput
            mode={'flat'}
            keyboardType={'numeric'}
            style={styles.textInput}
            label={'Response Code'}
            onChangeText={setCode}
            value={code}
          />
          <TextInput
            mode={'flat'}
            keyboardType={'numeric'}
            style={styles.textInput}
            label={'Timeout value'}
            onChangeText={setTo}
            value={to}
          />
          <TextInput
            mode={'flat'}
            style={styles.textInput}
            label={'Response Method'}
            onChangeText={setMethod}
            value={method}
          />
          <TextInput
            mode={'flat'}
            style={[styles.textInput, styles.multiLine]}
            value={headers}
            label="Headers"
            multiline
            onChangeText={setHeaders}
          />
          <TextInput
            mode={'flat'}
            style={[styles.textInput, styles.multiLine]}
            label="Body"
            multiline
            onChangeText={setBody}
            value={body}
          />
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
  scrollView: {
    paddingTop: 26,
    paddingBottom: 20,
  },
  textInput: {
    margin: 10,
  },
  multiLine: {
    height: 200,
  },
});
