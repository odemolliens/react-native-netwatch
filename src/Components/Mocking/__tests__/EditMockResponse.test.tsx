import React, { useState } from 'react';
import { EditMockResponse } from '../EditMockResponse';
import { getMockResponses, mockRequestWithResponse, MockResponse } from '../utils';
import { shallow, ShallowWrapper } from 'enzyme';
import { NavBar } from '../../NavBar';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: (f: () => void) => f(),
}));

describe('Test Edit Mock Response Screen', () => {
  let component: ShallowWrapper;
  const mockResponse: MockResponse = {
    url: 'https://www.google.com',
    method: 'GET',
    headers: '',
    response: '',
    statusCode: 0,
    timeout: 0,
    active: true,
  };

  it('should render properly', () => {
    givenMockResponse();
    component = shallow(<EditMockResponse mockResponse={mockResponse} onPressBack={() => {}} update={false} />);
    expect(component).toMatchSnapshot();
  });

  it('should render properly when updating mock', () => {
    givenMockResponse();
    component = shallow(<EditMockResponse mockResponse={mockResponse} onPressBack={() => {}} update={true} />);
    expect(component).toMatchSnapshot();
  });

  it('should save mock response when pressing save button', function () {
    givenMockResponse();
    component = shallow(<EditMockResponse mockResponse={mockResponse} onPressBack={() => {}} update={false} />);
    // @ts-ignore
    component.find(NavBar).props().rightComponent?.props.onPress();
    expect(getMockResponses()[0]).toEqual(expect.objectContaining(mockResponse));
  });

  function givenMockResponse() {
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.url, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.method, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.headers, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.response, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.statusCode, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.active, jest.fn()]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse.timeout, jest.fn()]);
  }
});
