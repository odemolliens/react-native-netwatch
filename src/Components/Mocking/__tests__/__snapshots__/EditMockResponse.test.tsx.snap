// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Edit Mock Response Screen should render properly 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <NavBar
    onPressBack={[Function]}
    rightComponent={
      <Button
        onPress={[Function]}
        title="Save"
      />
    }
    title="Edit Response Mock"
  />
  <ScrollView
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingTop": 26,
      }
    }
    keyboardDismissMode="on-drag"
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "margin": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#F9FAFB",
            }
          }
        >
          Leave blank entries to keep unchanged.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#F9FAFB",
              "flex": 1,
              "fontSize": 14,
              "marginBottom": 10,
            }
          }
        >
          Mock response active
        </Text>
        <Switch
          onValueChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  true,
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          value={true}
        />
      </View>
      <Component
        label="URL"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "https://www.google.com",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value="https://www.google.com"
      />
      <Component
        keyboardType="numeric"
        label="Response Code"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "0",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value={0}
      />
      <Component
        keyboardType="numeric"
        label="Timeout value"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "0",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value={0}
      />
      <Component
        label="Response Method"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "GET",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value="GET"
      />
      <Component
        label="Headers"
        mode="flat"
        multiline={true}
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Array [
            Object {
              "margin": 10,
            },
            Object {
              "height": 200,
            },
          ]
        }
        value=""
      />
      <Component
        label="Body"
        mode="flat"
        multiline={true}
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Array [
            Object {
              "margin": 10,
            },
            Object {
              "height": 200,
            },
          ]
        }
        value=""
      />
    </View>
  </ScrollView>
</KeyboardAvoidingView>
`;

exports[`Test Edit Mock Response Screen should render properly when updating mock 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <NavBar
    onPressBack={[Function]}
    rightComponent={
      <Button
        onPress={[Function]}
        title="Save"
      />
    }
    title="Edit Response Mock"
  />
  <ScrollView
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingTop": 26,
      }
    }
    keyboardDismissMode="on-drag"
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "margin": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#F9FAFB",
            }
          }
        >
          Leave blank entries to keep unchanged.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#F9FAFB",
              "flex": 1,
              "fontSize": 14,
              "marginBottom": 10,
            }
          }
        >
          Mock response active
        </Text>
        <Switch
          onValueChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  true,
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          value={true}
        />
      </View>
      <Component
        label="URL"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "https://www.google.com",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value="https://www.google.com"
      />
      <Component
        keyboardType="numeric"
        label="Response Code"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "0",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value={0}
      />
      <Component
        keyboardType="numeric"
        label="Timeout value"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "0",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value={0}
      />
      <Component
        label="Response Method"
        mode="flat"
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "GET",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "margin": 10,
          }
        }
        value="GET"
      />
      <Component
        label="Headers"
        mode="flat"
        multiline={true}
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Array [
            Object {
              "margin": 10,
            },
            Object {
              "height": 200,
            },
          ]
        }
        value=""
      />
      <Component
        label="Body"
        mode="flat"
        multiline={true}
        onChangeText={
          [MockFunction] {
            "calls": Array [
              Array [
                "",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Array [
            Object {
              "margin": 10,
            },
            Object {
              "height": 200,
            },
          ]
        }
        value=""
      />
    </View>
  </ScrollView>
</KeyboardAvoidingView>
`;
